File: Inventarios.Api.Nest/productos.api/src/domain/common/result.ts
export class Result<T> {
  readonly isSuccess: boolean;
  readonly isFailure: boolean;
  readonly value?: T;
  readonly error?: string;

  private constructor(isSuccess: boolean, value?: T, error?: string) {
    this.isSuccess = isSuccess;
    this.isFailure = !isSuccess;
    this.value = value;
    this.error = error;
  }

  static success<T>(value: T): Result<T> {
    return new Result(true, value);
  }

  static failure<T>(error: string): Result<T> {
    return new Result<T>(false, undefined, error);
  }
}

export class VoidResult {
  readonly isSuccess: boolean;
  readonly isFailure: boolean;
  readonly error?: string;

  private constructor(isSuccess: boolean, error?: string) {
    this.isSuccess = isSuccess;
    this.isFailure = !isSuccess;
    this.error = error;
  }

  static success(): VoidResult {
    return new VoidResult(true);
  }

  static failure(error: string): VoidResult {
    return new VoidResult(false, error);
  }
}

File: Inventarios.Api/Inventarios.api.Domain/Common/Result.cs
namespace Inventarios.api.Domain.Common;

/// Result Pattern: evita lanzar excepciones para flujos de negocio normales.

public class Result<T>
{
    public bool IsSuccess { get; }
    public T? Value { get; }
    public string? Error { get; }
    public bool IsFailure => !IsSuccess;

    protected Result(bool isSuccess, T? value, string? error)
    {
        IsSuccess = isSuccess;
        Value = value;
        Error = error;
    }

    public static Result<T> Success(T value) => new(true, value, null);
    public static Result<T> Failure(string error) => new(false, default, error);
}

public class Result
{
    public bool IsSuccess { get; }
    public string? Error { get; }
    public bool IsFailure => !IsSuccess;

    protected Result(bool isSuccess, string? error)
    {
        IsSuccess = isSuccess;
        Error = error;
    }

    public static Result Success() => new(true, null);
    public static Result Failure(string error) => new(false, error);
}